src/qt/netwatch.cpp

```diff
@@ -0,0 +1,798 @@
+            // Check for a match
+            fAdding = newsearch->match(*this, row);
+        }
+        if (fAdding) {
+            results.append(row);
+        }
+    }
+}
+
+void NetWatchLogModel::searchDisable() {
+    LOCK(cs);
+    delete currentSearch;
+    currentSearch = NULL;
+}
+
+void NetWatchLogModel::log_append(const LogEntry& le, size_t& rows_used)
+{
+    AssertLockHeld(cs);
+    if (logskip) {
+        // Replace a deleted row
+        getLogEntryRow(rows_used) = le;
+        --logskip;
+    } else {
+        // Haven't filled up yet, so just push_back
+        assert(!logpos);
+        log.push_back(le);
+    }
+    ++rows_used;
+    if (rows_used > max_nonweak_txouts) {
+        LogEntry& old_le = getLogEntryRow(rows_used - max_nonweak_txouts - 1);
+        if (old_le) {
+            old_le.makeWeak();
+        }
+    }
+}
+
+void NetWatchLogModel::LogAddEntry(const LogEntry& le, size_t vout_count)
+{
+    if (vout_count < 1) {
+        vout_count = 1;
+    }
+    const QModelIndex dummy;
+    LOCK(cs);
+    size_t rows_to_remove = 0;
+    if (vout_count >= max_vout_per_tx) {
+        vout_count = max_vout_per_tx;
+    }
+    size_t rows_used = rowCount();
+    if (rows_used > logsizelimit - vout_count) {
+        rows_to_remove = (rows_used + vout_count) - logsizelimit;
+    }
+    if (rows_to_remove) {
+        // Don't orphan continuation entries
+        while (isLogRowContinuation(rows_to_remove)) {
+            ++rows_to_remove;
+        }
+
+        beginRemoveRows(dummy, 0, rows_to_remove - 1);
+        logpos = (logpos + rows_to_remove) % log.size();
+        logskip += rows_to_remove;
+        endRemoveRows();
+
+        rows_used = rowCount();
+    }
+
+    const LogEntry cont_le;
+    const int first_new_row = rows_used, last_new_row = rows_used + vout_count - 1;
+    beginInsertRows(dummy, first_new_row, last_new_row);
+    log_append(le, rows_used);
+    for (size_t i = 1; i < vout_count; ++i) {
+        log_append(cont_le, rows_used);
+    }
+    endInsertRows();
+
+    if (currentSearch) {
+        QList<int> new_matches;
+        if (currentSearch->fCheckAddr || currentSearch->fCheckValue) {
+            for (int row = first_new_row; row <= last_new_row; ++row) {
+                if (currentSearch->match(*this, row)) {
+                    new_matches.append(row);
+                }
+            }
+        } else if (currentSearch->match(*this, first_new_row)) {
+            for (int row = first_new_row; row <= last_new_row; ++row) {
+                new_matches.append(row);
+            }
+        }
+        if (!new_matches.isEmpty()) {
+            Q_EMIT moreSearchResults(new_matches);
+        }
+    }
+}
+
+void NetWatchLogModel::LogBlock(const CBlockIndex* pblockindex)
+{
+    CBlock block;
+    if (!ReadBlockFromDisk(block, pblockindex, Params().GetConsensus())) {
+        // Indicate error somehow?
+        return;
+    }
+    assert(block.vtx.size());
```

Comment:
What we probably should do is adding each blocks size and vtx.size() to CBlockIndex* (would require to alter the block index, migration, etc. yes).